assembler.py

```diff
@@ -55,15 +55,43 @@ def read_source(self, filename=None):
                 f'Unsupported file type [{self.filename}]. Only accepting files ending in .asm')
         source = open(self.filename, 'r')
         lines = source.readlines()
-        for line in lines:
-            if line != '\n':
-                self.micro_instr.append(line.strip())
+        if self.is_indented(lines[0]):
+            raise AssertionError('Indentation Error: the first line cannot be indented.')
+        else:
+            self.verify_indentation(lines[0], 0)
+            self.micro_instr.append(lines[0].strip())
```

Comment:
Why do you check line[0] twice?

Reply:
The verify_indentation method checks for three things: 

- whether the line has a tab 
- if a line is indented with the wrong amount of spaces 
- if a line with ':' is indented 

The first line can pass all those checks, but the first line can't be indented under any circumstance, causing the need for another check. In short, it checks for indentation and it checks for the other three cases as well

Reply:
Why not just add that extra check you are doing inside of `verify_indentation` method?

Reply:
I just saw that I added the index as a parameter for that method. I used to have it like that because the method couldn't find a way to know if that line being verified was the first one. I'll include it into the method now.